text/0000-try-trait-v2.md

```diff
@@ -0,0 +1,837 @@
+    }
+});
+assert!(matches!(x, MyResult::Terrific(Some(2))));
+```
+
+As expected, the mixing in `bar` no longer compiles:
+```
+help: the trait `Try<std::result::Result<!, {float}>>` is not implemented for `MyResult<(), f32>`
+```
+
+### Enabling `Result`-like error conversion
+
+`Result` allows mismatched error types so long as it can convert the source one into the type on the function.  But if we try that with our current type, it won't work:
+```rust
+fn qux() -> MyResult<(), i64> {
+    let _: () = MyResult::Unfortunate(3_u8)?;
+    MyResult::Terrific(())
+}
+```
+
+That help message in the error from the previous section gives us a clue, however.  Thus far we've been using the default generic parameter on `Try<H>` which only allows an exact match to our declared `Bubble::Holder` type.  But we can be more general if we want.  Let's use `From`, like `Result` does:
+```rust
+impl<T, U, V: From<U>> Try<MyResult<Infallible, U>> for MyResult<T, V> {
+    fn from_holder(h: MyResult<Infallible, U>) -> Self {
+        match h {
+            MyResult::Unfortunate(e) => MyResult::Unfortunate(From::from(e)),
+            MyResult::Terrific(v) => match v {},
+        }
+    }
+}
+```
+
+With that the `qux` example starts compiling successfully.
+
+(This can also be used to allow interconversion with holder types from other type constructors, but that's left as an exercise for the reader.)
+
+<!--
+Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+-->
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `ops::ControlFlow`
+
+```rust
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum ControlFlow<B, C = ()> {
+    /// Exit the loop, yielding the given value
+    Break(B),
+    /// Continue in the loop, using the given value for the next iteration
+    Continue(C),
+}
+```
+
+## The traits
+
+```rust
+trait Bubble {
+	type Continue;
+	type Holder: BreakHolder<Self::Continue>;
+	fn continue_with(c: Self::Continue) -> Self;
+	fn branch(self) -> ControlFlow<Self::Holder, Self::Continue>;
+}
+
+trait BreakHolder<T> {
```

Comment:
```suggestion
trait BreakHolder<T>: Sized {
```
Without something like this it won’t compile. And it looks like your [own implementation](https://github.com/scottmcm/rust/blob/do-or-do-not/library/core/src/ops/try.rs#L132) has a bound like this, too. (An alternative to make the code compile is to relax the parameter `H` of `Try` to be `?Sized`.)